js/data/buffer.js

```diff
@@ -0,0 +1,335 @@
+
+        // Normalize the attributes
+        for (var key in options.attributes) {
+            var attribute = util.extend({}, options.attributes[key]);
+
+            attribute.name = attribute.name || key;
+            attribute.components = attribute.components || 1;
+            attribute.type = attribute.type || Buffer.AttributeType.UNSIGNED_BYTE;
+            attribute.size = attribute.type.size * attribute.components;
+            attribute.offset = this.itemSize;
+
+            this.itemSize = align(attribute.offset + attribute.size, attributeAlignment);
+
+            this.attributes[attribute.name] = attribute;
+        }
+    }
+
+    // Enable some shortcuts if there is only one attribute on this buffer.
+    var attributeNames = Object.keys(this.attributes);
+    if (attributeNames.length === 1) {
+        this.singleAttribute = this.attributes[attributeNames[0]];
+    } else {
+        this.singleAttribute = null;
+    }
+}
+
+// Start Buffer1 compatability code
+Buffer.prototype = {
+
+    get pos() { return this.length * this.itemSize; },
+    get index() { return this.length; },
+
+    get arrayType() { return this.type; },
+
+    get array() { return this.arrayBuffer; },
+    set array(value) {
+        assert(value === null);
+        this.arrayBuffer = null;
+    },
+
+    // binds the buffer to a webgl context
+    bind: function(gl) {
+        var type = gl[this.arrayType];
+        if (!this.buffer) {
+            this.buffer = gl.createBuffer();
+            gl.bindBuffer(type, this.buffer);
+            gl.bufferData(type, this.array.slice(0, this.pos), gl.STATIC_DRAW);
+
+            // dump array buffer once it's bound to gl
+            this.array = null;
+        } else {
+            gl.bindBuffer(type, this.buffer);
+        }
+    },
+
+    destroy: function(gl) {
+        if (this.buffer) {
+            gl.deleteBuffer(this.buffer);
+        }
+    },
+
+    setupViews: function() { /* intentional no-op */ }
+
+};
+// End Buffer1 compatability code
+
+/**
+ * Push an item onto the end of the buffer. Grows the buffer if necessary.
+ * @private
+ * @param {(BufferItem|BufferValue)} item Item to be appended. If the buffer only has one attribute,
+ * can be a single value.
+ * @returns {number} The index of the appended item.
+ */
+Buffer.prototype.push = function(item) {
+    var index = this.length;
+    this.length++;
+
+    if (this.getIndexOffset(this.length) > this.capacity) {
+        this._resize(this.capacity * 1.5);
+    }
+
+    this.set(index, item);
+
+    return index;
+};
+
+/**
+ * Set an item at a particular index. Grows the buffer if necessary.
+ * @private
+ * @param {number} index The index of the item to set
+ * @param {(BufferItem|BufferValue)} item the item to set. If the buffer only has one attribute,
+ * it can be a single value instead of an item.
+ */
+Buffer.prototype.set = function(index, item) {
+    if (!Array.isArray(item)) {
```

Comment:
Performance optimization: eliminate this condition. Either:

* Always require an array, or
* Write `this.set = function() { ... }` dynamically in the constructor, and make it require an array for multiple attributes, and require no array for single attributes. For multiple attributes, it may help to iterate over `options.attributes` rather than `item` (like, maybe the compiler will detect that `options.attributes` is constant).

Writing `this.set = function() { ... }` dynamically in the constructor can hopefully give you some nice speed ups, while stopping just short of needing code-gen + `eval`.